packages/vuetify/src/composables/locale.ts

```diff
@@ -0,0 +1,166 @@
+import { computed, inject, isRef, provide, ref } from 'vue'
+import { getObjectValueByPath, consoleError, consoleWarn } from '@/util'
+
+import en from '@/locale/en'
+
+// Types
+import type { App, InjectionKey, Ref } from 'vue'
+
+export interface LocaleOptions {
+  defaultLocale?: string
+  fallbackLocale?: string
+  messages?: Record<string, any>
+}
+
+export interface LocaleProps {
+  locale?: string
+  fallbackLocale?: string
+  messages?: Record<string, any>
+}
+
+export interface LocaleInstance {
+  locale: Ref<string>
+  fallbackLocale: Ref<string>
+  messages: Ref<Record<string, any>>
+  t: (key: string, ...params: unknown[]) => string
+}
+
+export interface LocaleAdapter {
+  createRoot: (app: App) => LocaleInstance
+  getScope: () => LocaleInstance
+  createScope: (options?: LocaleProps) => LocaleInstance
+  rtl?: Record<string, boolean>
+}
+
+export const VuetifyLocaleAdapterSymbol: InjectionKey<LocaleAdapter> = Symbol.for('vuetify:locale-adapter')
+export const VuetifyLocaleSymbol: InjectionKey<LocaleInstance> = Symbol.for('vuetify:locale')
+
+export function provideLocale (props?: LocaleProps) {
+  const adapter = inject(VuetifyLocaleAdapterSymbol)
+
+  if (!adapter) throw new Error('Could not find injected locale adapter')
+
+  return adapter.createScope(props)
+}
+
+export function useLocale () {
+  const adapter = inject(VuetifyLocaleAdapterSymbol)
+
+  if (!adapter) throw new Error('Could not find injected locale adapter')
+
+  return adapter.getScope()
+}
+
+function isLocaleAdapter (x: any): x is LocaleAdapter {
+  return !!x && x.hasOwnProperty('getScope') && x.hasOwnProperty('createScope') && x.hasOwnProperty('createRoot')
+}
+
+export function createLocaleAdapter (app: App, options?: LocaleOptions | LocaleAdapter) {
+  const adapter = isLocaleAdapter(options) ? options : createDefaultLocaleAdapter(options)
+
+  const rootInstance = adapter.createRoot(app)
+
+  return { adapter, rootInstance }
+}
+
+const LANG_PREFIX = '$vuetify.'
```

Comment:
I did this earlier, but realized that it doesn't really work for us. It would mean that if you use vue-i18n integration, the messages object would look like this:

```js
{
  foo: 'bar',
  $badge: '...',
  $dataIterator: { ... },
  ...
}
```

instead of

```js
{
  foo: 'bar':
  $vuetify: {
    badge: ...,
    dataIterator: ...
  },
  ...
}
```